Add functionality to top off battery without changing underlying BPM mode.

DimseBoms opened the pull request:
Added functionality to temporarily charge the battery to 100% without changing the underlying BPM mode. Quite useful if you mostly work at home but need a full charge once in a while. The old BPM mode will be reinstated if the charger is disconnected or if the battery completes a full charge.

The function is only available if the charger is connected and BPM is not disabled.

Since it reads values directly from /sys and /etc it's going to know if the extension has been restarted and it will restart the battery watcher if it notices that the Top_off function is supposed to be active.

![bilde](https://user-images.githubusercontent.com/72513291/175818830-defb4335-adcf-4f95-a878-bbe429b35f91.png)

Commit:
Added function to charge battery to 100% once

DimseBoms committed:
Merge branch 'egormanga:master' into master

Commit:
Added comments

Commit:
e

Commit:
Added chargin check on handle state change

egormanga commented on extension.js:
```diff
@@ -154,17 +168,14 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
 			try {
-				file_sys.replace_contents(`${low} ${high}`, null, false, 0, null);
-				file_def.replace_contents(`${low} ${high}`, null, false, 0, null);
+				this._file_sys.replace_contents(`${low} ${high}`, null, false, 0, null);
```
Would this still work after the module reinserted without reopening the file?

DimseBoms replied:
I don't know what you mean about reloading the module but the files stays open as instance variables on line 43 and 44. It is done this way since there is now multiple methods where the files are going to be used.

One issue that I see now is that the If on 160/171 is not wrapped in a block..

How would I go about editing these requested changes. Do I just commit them to my fork?

egormanga replied:
I mean `rmmod`'ing then `modprobe`'ing the [kernel module](https://github.com/aymanbagabas/Huawei-WMI), e.g. during its development. The file descriptor will then become stale and unusable, that's the exact reason why I reopen them each time in the first place.

The `if` on L171 does not need its content wrapped, as it's a single statement (a `try..catch` block). Per my code style, the braces are redundant in this case. Please remove them, too.

And yes, you should just commit and push your changes to the same branch — GitHub will handle the rest.

DimseBoms replied:
If that is the case then I'm going to rework the code to use the previous method of loading files. I did not even realise that this could be an issue as I'm still a novice. Thank you for pointing this out :)

egormanga replied:
Okay, that's kind of correct now. I'll probably rework variable naming later; good for now.

egormanga commented on extension.js:
```diff
@@ -174,6 +185,98 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
 
+	_update_top_off() {
+		this._get_battery(proxy => {
+			let is_discharging = proxy.State === UPower.DeviceState.DISCHARGING
+			let is_fully_charged = proxy.State === UPower.DeviceState.FULLY_CHARGED
```
Code style: suggest adding parentheses around the comparisons.

egormanga commented on extension.js:
```diff
@@ -174,6 +185,98 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
+
+	_start_top_off() {
+		this._get_battery(proxy => {  // Connects watcher
+			this._battery_watching = proxy.connect(
+				'g-properties-changed',
+				this._update_top_off.bind(this)
+			);
```
Code style: I'd like it more non-line-wrapped.

egormanga commented on extension.js:
```diff
@@ -174,6 +185,98 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
+		let def_low, def_high;
+		this._get_battery(proxy => {  // Disconnects watcher
+			proxy.disconnect(this._battery_watching);
+			try {	// Reinstates old BPM values
```
Code style: the comment here is separated with a tab instead of 2 spaces.

egormanga commented on extension.js:
```diff
@@ -174,6 +185,98 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
+				this._file_sys.replace_contents(`${def_low} ${def_high}`, null, false, 0, null);
+				this._topping_off = false;
+			} catch (e) {}
+			})
```
Code style: indentation?

egormanga commented on extension.js:
```diff
@@ -174,6 +185,98 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
+		
+		// Handle state change
+		if (state !== undefined)
+		try {
+			if (state && !this._topping_off && is_charging) this._start_top_off();  // Top off switch gets switched on
+			else if (!state && this._topping_off) this._stop_top_off(); 		    // Top off switch gets switched off
+		} catch (e) {
+			global.log(e)
+		}
```
Code style: indentation; 2 spaces instead of a tab to separate the comments.

egormanga commented on extension.js:
```diff
@@ -174,6 +185,98 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
+				this._top_off.setToggleState(false);
+				this._top_off.setSensitive(false);
+			}
+            // If BPM IS on AND device is charging -> available
```
Code style: indentation (the code uses tabs, here are spaces).

egormanga commented on extension.js:
```diff
@@ -174,6 +185,98 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
+				let top_is_active = ((def_low != 0 && def_high != 100) && (sys_low == 0 && sys_high == 100));
+				if (top_is_active && is_charging) {
+					this._top_off.setToggleState(true);
+					if (top_is_active && !this._topping_off) this._start_top_off(); // Reconnects watcher if extension has been restarted without reinstating BPM
+				} // If top is active and charger is not connected -> reinstate old BPM
+				else if (top_is_active && !is_charging) {
+					this._stop_top_off();
+					this._top_off.setToggleState(false);
+					this._top_off.setSensitive(false);
+				} 
+				} else { // Reinstates old BPM in case of unclean watcher exit and handles edge cases
```
Code style: 2 spaces instead of 1 to separate comments.

egormanga commented on extension.js:
```diff
@@ -234,3 +337,4 @@ function init(meta) {
 
 // by Sdore, 2021-22
 //   apps.sdore.me
+
```
Code style: just why... please remove this line.

egormanga reviewed (changes requested):
LGTM otherwise. Will accept after the fixes. Thank you!

Commit:
Requested changes pull#4

egormanga commented on extension.js:
```diff
@@ -174,6 +186,95 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
+				this._file_sys.replace_contents(`${def_low} ${def_high}`, null, false, 0, null);
+				this._topping_off = false;
+			} catch (e) {}
+		})
```
There's a missing `;` in this line (to close `this._get_battery(`).

egormanga commented on extension.js:
```diff
@@ -233,4 +334,4 @@ function init(meta) {
 
 // by Sdore, 2021-22
-//   apps.sdore.me
+//   apps.sdore.me
```
Now, there should be a newline — just a single, not two as you did earlier. :)

Commit:
Reworked file loading, added semicolon PR#4

egormanga commented:
@sermart1234 Please test that PR on your device model.

DimseBoms committed:
Fix button availability bug

Fixed a bug where the button would be unavailable(not sensitive) if the charging treshold was met and charger was plugged in. Cleaned up some code and removed some code that was never ran.

DimseBoms committed:
Add top off button refresh after clicking

egormanga commented on extension.js:
```diff
@@ -283,6 +270,16 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
+
+		// Handle state change
+		if (state !== undefined)
+		try {
+			if (state && !this._topping_off) this._start_top_off();  // Top off switch gets switched on
+			else if (!state && this._topping_off) this._stop_top_off();  // Top off switch gets switched off
+			this._set_top_off();
+		} catch (e) {
+			global.log(e)
+		}
```
Please indent that block properly (+1 tab).

egormanga commented on extension.js:
```diff
@@ -236,43 +236,30 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
 					this._top_off.setToggleState(true);
 					if (top_is_active && !this._topping_off) this._start_top_off();  // Reconnects watcher if extension has been restarted without reinstating BPM
-				}  // If top is active and charger is not connected -> reinstate old BPM
-				else if (top_is_active && !is_charging) {
```
Please join these two lines, as per my code style.

egormanga commented on extension.js:
```diff
@@ -236,43 +236,30 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
-					this._top_off.setToggleState(false);
-					this._top_off.setSensitive(false);
-				}
+				}  // If top is not active -> Button = Off
+				else this._top_off.setToggleState(false);
```
Please join these two lines, as per my code style.

egormanga reviewed (changes requested):
Perfectly fine otherwise, will merge.

egormanga commented:
Sorry for the delay. @sermart1234 has become unavailable lately, so I'll have to test this myself, thus on my laptop model only.

DimseBoms committed:
Joined lines and increased indentation

egormanga commented on extension.js:
```diff
@@ -174,6 +188,99 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
+					this._top_off.setToggleState(true);
+					if (top_is_active && !this._topping_off) this._start_top_off();  // Reconnects watcher if extension has been restarted without reinstating BPM
+				} else this._top_off.setToggleState(false);  // If top is not active -> Button = Off
+				} else {  // Reinstates old BPM in case of unclean watcher exit and handles edge cases
+					if (def_low != sys_low || def_high != sys_high) this._stop_top_off();
+					this._top_off.setToggleState(false);
+					this._top_off.setSensitive(false);
+				}
```
This bit of code is not properly intended now. Should be 1 tab less.

Upd: or maybe I didn't notice it earlier.

egormanga commented on extension.js:
```diff
@@ -174,6 +188,99 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
+			if (def_low == 0 && def_high == 100) {
+				this._top_off.setToggleState(false);
+				this._top_off.setSensitive(false);
+			}
+			// If BPM is on and device is not discharging -> Button = Available
+			else if ((def_low != 0 || def_high != 100) && !is_discharging ) {
```
Remove the unnecessary space before the closing bracket at line 254 and merge these to a single line too.

In general, don't place an `else` clause with a line break after the previous `}` it corresponds to, cause that's hard to follow.

egormanga commented:
Actually, I should've merged this immediately and then fix the style manually. But now it's too late, so fine.

egormanga commented:
Yeah, I somehow forgot about this completely. Sorry for that.
